### PR TITLE
Add botframework attachements to tracker

### PR DESCRIPTION
**Pull request by 1tupac**
**Proposed changes**:
I need to be able to receive attachments sent through Microsoft Teams (BotFramework). That's why I made some modifications to add the attachement data to the tracker. 

Let me know what you think about it. 

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

### CODE REVIEW

**Comment by msamogh** [comment]
Thanks for the PR, we'll give it a review as soon as possible.

File: rasa/core/processor.py [hunk 1]
```diff
@@ -297,7 +297,9 @@ def _get_action(self, action_name):
     ) -> None:
 
         if message.parse_data:
+            parse_message = await self._parse_message(message)
             parse_data = message.parse_data
+            parse_data.update(parse_message)
```
Comment (ricwo) [inline rasa/core/processor.py]: why this change? how is it related to the changes to `botframework.py`?

**Comment by 1tupac** [comment]
Hi @ricwo, yes from May 19, Microsoft added the ability [to upload and download files](https://docs.microsoft.com/en-us/microsoftteams/platform/concepts/bots/bots-files#notifying-the-user-about-an-uploaded-file) in a bot conversation using BotFramework. I'm currently using it for my chatbot to retrieve file when necessary and I think it can be useful for the rest of Rasa community.

File: rasa/core/processor.py (hunk 1 above)
Comment (1tupac) [reply rasa/core/processor.py]: Because when I receive an attachement, I set `parse_data` with it in the UserMessage ([line 216](https://github.com/RasaHQ/rasa/pull/4135/files#diff-d27075473e03e27edac04ffe9aa72fb6R216)). 

However, the processor doesn't parse the message if `parse_data` contains a value, leading to an error [when accessing the keys](https://github.com/RasaHQ/rasa/blob/master/rasa/core/processor.py#L309) `intent` and `entities`. I chose to execute `_parse_message` even if the message contains an attachement because some text can be sent with the attachement and it's interesting to be able to parse it alongside the uploaded file.

Comment (1tupac) [reply rasa/core/processor.py]:
```suggestion
    attachment = postdata['attachments'][0]
    user_msg = UserMessage(
    text=postdata['text'] if postdata.get('text') else "",
    parse_data={
        "attachment": attachment,
        "intent": {"confidence": 1.0, "name": "attachment"},
        "entities": [],
        "text": postdata['text'] if postdata.get('text') else "",
    },
    output_channel=out_channel,
    sender_id=postdata["from"]["id"],
    input_channel=self.name(),
)
```
Otherwise we could do something like this. What do you think about it?

Comment (ricwo) [reply rasa/core/processor.py]: but the point of this check is to _not_ parse a message again that already contains `parse_data()`

File: rasa/core/channels/botframework.py [hunk 2]
```diff
@@ -208,13 +208,23 @@ def blueprint(self, on_new_message):
...
-                        out_channel,
-                        postdata["from"]["id"],
-                        input_channel=self.name(),
-                    )
+                    
+                    if postdata.get('attachments'):
+                        user_msg = UserMessage(
+                            text=(postdata['text'] if postdata.get('text') 
+                                  else ""),
+                            parse_data={"attachement": postdata['attachments'][0]},
```
Comment (ricwo) [inline rasa/core/channels/botframework.py]: I don't think `parse_data` is the right place for an attachment. it should contain the result of an intent prediction

Comment (1tupac) [reply rasa/core/channels/botframework.py]: You're right! Awesome that UserMessage now support metadata. It works like a charm!

File: rasa/core/processor.py (hunk 1 above)
Comment (1tupac) [reply rasa/core/processor.py]: Yup, fixed

File: rasa/core/channels/botframework.py [hunk 3]
```diff
@@ -208,13 +208,23 @@ def blueprint(self, on_new_message):
...
-                    user_msg = UserMessage(
-                        postdata["text"],
-                        out_channel,
-                        postdata["from"]["id"],
-                        input_channel=self.name(),
-                    )
+                    
+                    if postdata.get('attachments'):
+                        user_msg = UserMessage(
+                            text=(postdata['text'] if postdata.get('text')
```
Comment (ricwo) [inline rasa/core/channels/botframework.py]:
```suggestion
                            text=postdata.get("text", "")
```

Comment (ricwo) [reply rasa/core/channels/botframework.py]: I'm not sure I understand why this is necessary. what does it have to do with the presence of `postdata['attachments']`?

Comment (ricwo) [reply rasa/core/channels/botframework.py]: if not, we could keep the creation of `user_msg` as it is, and add the attachment later if necessary:
```python
if postdata.get('attachments'):
    user_msg.metadata = {"attachments": postdata['attachments']}
```

**Comment by tmbo** [comment]
@1tupac thanks for the PR - do you have time to take a look at the comments so we can get this merged?

File: rasa/core/channels/botframework.py (hunk 3 above)
Comment (1tupac) [reply rasa/core/channels/botframework.py]: The thing is sometimes, something can be uploaded in the chat and Microsoft removes the `text` key of the json, making `postdata['text']` throw an error...

File: CHANGELOG.rst [hunk 4]
```diff
@@ -27,6 +27,7 @@ Added
   (``rasa.core.agent.handle_channels()``). The number of workers can be set using the
   environment variable ``SANIC_WORKERS`` (default: 1). A value of >1 is allowed only in
   combination with ``RedisLockStore`` as the lock store.
+- Botframework channel can handle uploaded files in ``UserMessage`` metadata
```
Comment (ricwo) [inline CHANGELOG.rst]:
```suggestion
- Botframework channel can handle uploaded files in ``UserMessage`` metadata.
```

File: rasa/core/channels/botframework.py [hunk 5]
```diff
@@ -214,14 +214,19 @@ def blueprint(
...
-                        out_channel,
-                        postdata["from"]["id"],
+                        text=postdata.get("text", ""),
+                        output_channel=out_channel,
+                        sender_id=postdata["from"]["id"],
                         input_channel=self.name(),
                         metadata=metadata,
                     )
+                    if postdata.get('attachments'):
+                        user_msg.metadata = {
```
Comment (ricwo) [inline rasa/core/channels/botframework.py]: this would make the `self.get_metadata()` call in l. 206 useless. can we instead _add_ the `attachments` key to `metadata` if it's not `None`?

Comment (1tupac) [reply rasa/core/channels/botframework.py]: You're right! I moved the attachements detection in a function, let me know what you think about this

File: rasa/core/channels/botframework.py [hunk 6]
```diff
@@ -189,6 +189,23 @@ def __init__(self, app_id: Text, app_password: Text) -> None:
         self.app_id = app_id
         self.app_password = app_password
 
+    @staticmethod
+    def process_attachments(
```
Comment (ricwo) [inline rasa/core/channels/botframework.py]: how about something like 
```suggestion
    def add_attachments_to_metadata(
```

File: rasa/core/channels/botframework.py [hunk 7]
```diff
@@ -189,6 +189,23 @@ def __init__(self, app_id: Text, app_password: Text) -> None:
...
+        postdata: Dict[Text, Any], 
+        metadata: Optional[Dict[Text, Any]]
+    ) -> Optional[Dict[Text, Any]]:
```
Comment (ricwo) [inline rasa/core/channels/botframework.py]: Please add a docstring for this new method

File: rasa/core/channels/botframework.py [hunk 8]
```diff
@@ -205,6 +222,8 @@ def blueprint(
             postdata = request.json
             metadata = self.get_metadata(request)
 
+            metadata = self.process_attachments(postdata, metadata)
```
Comment (ricwo) [inline rasa/core/channels/botframework.py]:
```suggestion
            metadata_with_attachments = self.add_attachments_to_metadata(postdata, metadata)
```

File: rasa/core/channels/botframework.py [hunk 9]
```diff
@@ -214,14 +233,15 @@ def blueprint(
...
+                    
                     user_msg = UserMessage(
-                        postdata["text"],
-                        out_channel,
-                        postdata["from"]["id"],
+                        text=postdata.get("text", ""),
+                        output_channel=out_channel,
+                        sender_id=postdata["from"]["id"],
                         input_channel=self.name(),
                         metadata=metadata,
```
Comment (ricwo) [inline rasa/core/channels/botframework.py]:
```suggestion
                        metadata=metadata_with_attachments,
```

File: tests/core/test_channels.py [hunk 10]
```diff
@@ -417,6 +417,49 @@ def test_socketio_channel():
         assert text["text"] == "Hi there!"
 
 
+def test_botframework_attachments():
```
Comment (ricwo) [inline tests/core/test_channels.py]: nice test! can you also add test for the case where you supply a `metadata` dict?

Comment (1tupac) [reply tests/core/test_channels.py]: I've added a test, let me know what you think about it

Comment (ricwo) [reply tests/core/test_channels.py]: looks great 👍

File: rasa/core/channels/botframework.py [hunk 11]
```diff
@@ -189,6 +189,25 @@ def __init__(self, app_id: Text, app_password: Text) -> None:
         self.app_id = app_id
         self.app_password = app_password
 
+    @staticmethod
+    def add_attachments_to_metadata(
+        postdata: Dict[Text, Any], 
+        metadata: Optional[Dict[Text, Any]]
+    ) -> Optional[Dict[Text, Any]]:
+        """ Parse the message for the attachments values to add it to metadata
+        """
```
Comment (ricwo) [inline rasa/core/channels/botframework.py]:
```suggestion
```

File: rasa/core/channels/botframework.py [hunk 12]
```diff
@@ -189,6 +189,25 @@ def __init__(self, app_id: Text, app_password: Text) -> None:
         self.app_id = app_id
         self.app_password = app_password
 
+    @staticmethod
+    def add_attachments_to_metadata(
+        postdata: Dict[Text, Any], 
+        metadata: Optional[Dict[Text, Any]]
+    ) -> Optional[Dict[Text, Any]]:
+        """ Parse the message for the attachments values to add it to metadata
```
Comment (ricwo) [inline rasa/core/channels/botframework.py]: This sounds a little confusing - can we find a better way of phrasing this? How about something like

 ```suggestion
        """Merge the values of `postdata['attachments']` with `metadata`."""
```

**Comment by ricwo** [comment]
@1tupac would you mind fixing the failing test? https://travis-ci.com/RasaHQ/rasa/jobs/246675675#L1198

**Comment by tmbo** [comment]
fixed and merged as part of #4638
